Improve CTE materialization execution tests

misterjpapa opened the pull request:
Add method verifyResults and checkCTEInfoMatch to test whether CTEs are materialized by checking if the Explain plan has the expected CTEInfo element.

Existing tests were only asserting if results of executing the query are the same with and without using a CTE materialization strategy. These new methods go further and also check that the expected CTEs were materialized.

Refactor existing tests to use the verifyResults method or in the case of more complex queries the method checkCTEInfoMatch. In order for Explain plan to include the CTEInfo element, the system property VERBOSE_OPTIMIZER_INFO_ENABLED needs to be set for the session.

## Description
The existing CTE execution tests are only checking that the query results are the same with and without using a CTE materialization strategy during execution. Two new test helper methods verifyResults and checkCTEInfoMatch were introduced to also check the the Explain plan for the query execution included a CTEInfo element to verify that the CTEs are materialized.  

## Motivation and Context
Fix an open issue about improving existing CTE materialization tests to check beyond just the results of the query and verify the CTE materialization. See https://github.com/prestodb/presto/issues/21688

## Impact
The only changes are to improve existing tests. There is no public facing functionality impact.

## Test Plan
The only changes are to improve existing tests.

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
```
== NO RELEASE NOTE ==
```

misterjpapa commented:
Looking at the ci/circleci linux-presto-e2e-tests errors.

elharo commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -60,45 +68,41 @@ protected QueryRunner createQueryRunner()
-                        sql),
-                queryRunner.execute(getSession(),
-                        sql));
+        CTEInfoValue t = new CTEInfoValue("t", 1, false);
```
can we have less abbreviated variable names?

misterjpapa replied:
I've moved the CTEInfoValue objects creation inline with the ImmutableList similar to the paradigm in other tests which circumvents the need to define variable names.

elharo commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -60,45 +68,41 @@ protected QueryRunner createQueryRunner()
     public void testCteExecutionWhereOneCteRemovedBySimplifyEmptyInputRule()
     {
-        String sql = "WITH t as(select orderkey, count(*) as count from (select orderkey from orders where false) group by orderkey)," +
+        @Language("SQL") String sql = "WITH t as(select orderkey, count(*) as count from (select orderkey from orders where false) group by orderkey)," +
```
I know we use this Language annotation in various places, but honestly I'm not a big fan of it. If I had my druthers, I'd drop it.

misterjpapa replied:
I've removed the Language annotation as it wasn't originally used in this class. If someone objects my suggestion would be to open a new issue so we can decide then.

elharo commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -1242,14 +1261,28 @@ private void compareResults(MaterializedResult actual, MaterializedResult expect
+    private void checkCTEInfoMatch(String explain, List<CTEInfoValue> expectedCTEInfoValues)
+    {
+        String regex = "CTEInfo.*";
+        Pattern pattern = Pattern.compile(regex);
```
move this to a constant to avoid recomputing

misterjpapa replied:
Good idea - since latest changes.

jaystarshot replied:
https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md
our developer guide does seem to prefer immutable list

jaystarshot commented:
Thanks for your contribution! 
> Refactor the rest of the tests (45 remaining tests)
Wondering if there was any blocker for these, if not can we include them in the same commit?

misterjpapa commented:
No blocker Jay. I didn't want to change all 55 tests without getting confirmation from the community that I was heading in the right direction. I will address elharo's comments above and refactor the rest of the tests in the same commit.  With regards to elharo's comment above about "Language annotation in various places" for consistently sake should I just revert the annotation additions for the 10 tests so far or continue to add the annotation to the rest of the test methods? I am also not a big fan of the Language annotation, but it appeared to be part of the convention/warning for this codebase so I decided to follow it.

elharo commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -65,10 +74,8 @@ public void testCteExecutionWhereOneCteRemovedBySimplifyEmptyInputRule()
-                        sql),
-                queryRunner.execute(getSession(),
-                        sql));
+        verifyResults(queryRunner, sql, ImmutableList.of(new CTEInfoValue("t", 1, false),
```
doesn't matter a great deal but I tend to prefer Arrays.asList to ImmutableList.of to minimize thrid party dependencies

elharo replied:
"Prefer Immutable collections in Guava when possible." My experience is that it's not possible or advisable as much as most people think. I've had too many really ugly debugging sessions in production systems because an immutable type snuck into some code that expected to be able to mutate it. The root problem is that the Java Collections API was misdesigned 25 years ago in a way that violates the Liskov substitution principle, but we can't fix that here so the best we can do is not use types that don't implement the full List API. If we do want to use immutable types, we should make sure the API reflects that so folks aren't surprised when it doesn't mutate, but then we tightly couple to a third party API, which we also don't want to do. There's no good answer here.

All that being said, it's also possible to use Collections.unmodifiableList(Arrays.asList(...)) and retain immutability while avoiding third party dependencies. In this case since it's test code I'd prefer to keep it smaller as the immutability really isn't needed, but up to you.

jaystarshot replied:
Sg, maybe we should update the guide

misterjpapa replied:
+1 on "really ugly debugging sessions in production systems because an immutable type snuck into some code that expected to be able to mutate it", but as Jay mentioned it's explicitly called out in the Developer Guide that Immutable types should be used. Perhaps suggest an update to the Developer Guide for more nuisance use of the Immutable collections if we can come to a consensus.

jaystarshot commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -1261,9 +1257,24 @@ protected Session getSession()
+    public static class CTEInfoValue
+    {
+        String name;
+        int frequency;
```
Can you please add a materialized boolean here? Just in case we might want to assert non materialized ctes

jaystarshot commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -1261,9 +1257,24 @@ protected Session getSession()
                 .build();
     }
+
+    public static class CTEInfoValue
```
Also can we just reuse the CTEInformation class from the session?

misterjpapa replied:
Good point. It didn't register in my mind to use the CTEInformation class. I can created a CTEInformation rather than a CTEInfoValue, but I won't have access to a cteId value so it will stay null or match the name. Is that acceptable - otherwise I will just continue to use the CTEInfoValue inner class and extend it with the isMaterialized boolean? As mentioned in the earlier comment, it would have been great to just use the CTEInformation from the Session object directly, but that Session is created in-line when the DispatchManager is creating an internal query which the tests doesn't have access to.

jaystarshot replied:
> I won't have access to a cteId value so it will stay null or match the name.
Yes same is fine.

jaystarshot commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -1227,10 +1187,34 @@ public void testNestedCteWithSameName()
+
+    private void verifyResults(Session materializedSession, Session session, QueryRunner queryRunner, String query, List<CTEInfoValue> expectedCTEInfoValues)
+    {
+        compareResults(queryRunner.execute(materializedSession,
```
Some tests use compareResuts with verify ordering set to true, so you might need that info

misterjpapa replied:
I noticed there isn't any place which calls the compareResults with the checkOrdering field as true yet, which is why I didn't add the boolean checkOrdering field to verifyResults per YAGNI principles. If there are some tests that should be verifying the order, I didn't see them, but I'm okay adding the field for future proofing of the verifyResults method if that is what you want.
<img width="1371" alt="Screenshot 2024-07-23 at 8 09 57 AM" src="https://github.com/user-attachments/assets/a34f37a8-39fd-4881-a1c0-dd9f767ffd5b">

jaystarshot replied:
Hmm thats strange, There are a couple of tests which use it https://github.com/prestodb/presto/blob/master/presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java#L677

jaystarshot replied:
Ohh just this one^

misterjpapa replied:
I see it now. Must have missed one. I will make the changes then.

jaystarshot commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -1242,14 +1230,26 @@ private void compareResults(MaterializedResult actual, MaterializedResult expect
+
+        String cteInfo = matcher.group();
+        for (CTEInformation value : expectedCTEInfoValues) {
+            assertTrue(cteInfo.contains(value.getCteName() + ": " + value.getNumberOfReferences() + " (is_view: " + value.getIsView() + ")"),
```
Need to check the is materialized

misterjpapa replied:
The Explain plan doesn't contain information if the CTE is materialized. It is of the form "CTEInfo: [temp: 2 (is_view: false)]". Is there some other way to check?

jaystarshot replied:
Seems like a bug, let me see if we can fix this

jaystarshot replied:
https://github.com/prestodb/presto/pull/23288

misterjpapa replied:
That was quick. I wait for your changes in Issue #23288 to be merged and then add the same materialized check in these tests.

misterjpapa replied:
@jaystarshot - I noticed they merge your commit today. I will update the tests to take into account isMaterialized.

misterjpapa replied:
Updated the verifyResults method to check that the expected is_materialized value appears in the verbose Explain plan.

jaystarshot commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -1261,9 +1261,15 @@ protected Session getSession()
                 .build();
     }
+
+    private CTEInformation generateCTEInformation(String name, int frequency, boolean isView)
```
Nit: Maybe rename to `generateMaterializedCTEInformation`

misterjpapa commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -65,10 +75,8 @@ public void testCteExecutionWhereOneCteRemovedBySimplifyEmptyInputRule()
-                        sql),
-                queryRunner.execute(getSession(),
-                        sql));
+        verifyResults(queryRunner, sql, ImmutableList.of(generateMaterializedCTEInformation("t", 1, false, false),
```
@jaystarshot - please note that the is_materialized value for the cte named t is false.

misterjpapa commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -169,14 +168,18 @@ public void testComplexCommonFilterPushdown()
+                .setSystemProperty(CTE_MATERIALIZATION_STRATEGY, "HEURISTIC_COMPLEX_QUERIES_ONLY")
+                .setSystemProperty(CTE_FILTER_AND_PROJECTION_PUSHDOWN_ENABLED, "true")
+                .build();
+        verifyResults(materializedSession, getSession(), queryRunner, testQuery,
```
@jaystarshot please note that only the is_materialized value for the cte named "order_platform_data" is false, while the other 4 cte ("urgent", "medium", "high", "low") are all false.

jaystarshot commented:
Can you please squash all commits into one and clean up the commit message? Commit [standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
In presto we regenerally rebase and merge your commits of the pull request

misterjpapa commented:
@jaystarshot  - any thoughts why this test would be failing like so (https://github.com/prestodb/presto/actions/runs/10121353652/job/28061341426)? From the output it looks like a call  setSystemProperty(VERBOSE_OPTIMIZER_INFO_ENABLED, "true") against a session "bled" into the session that was used for the generation of the plan so the output is now verbose. 
<img width="1215" alt="Screenshot 2024-07-29 at 4 26 33 PM" src="https://github.com/user-attachments/assets/463bee83-389d-43ae-a894-a92ceca6691c"> resulting in a failure of comparison. Is that possible?

jaystarshot commented:
seems to be flaky, i reran and it was successful

misterjpapa commented:
> Can you please squash all commits into one and clean up the commit message? Commit [standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards). In presto we regenerally rebase and merge your commits of the pull request

I've rebased, squashed all the commits into one pull request, and have updated the commit message.

jaystarshot commented:
Still seeing 2 commits https://github.com/prestodb/presto/pull/23268/commits.

misterjpapa commented:
> Still seeing 2 commits https://github.com/prestodb/presto/pull/23268/commits.

I've squashed it to one commit now.

jaystarshot commented:
https://github.com/prestodb/presto/compare/a774895a44e7a020c2ee0b1cd0f30d56fd53e715..f1014ac79428dad93a528ad6c13288824f1f01fb
Your code regressed to the older version with CteInfoValue.

misterjpapa commented:
> https://github.com/prestodb/presto/compare/a774895a44e7a020c2ee0b1cd0f30d56fd53e715..f1014ac79428dad93a528ad6c13288824f1f01fb Your code regressed to the older version with CteInfoValue.

I think I've sorted out the squash issues. Apologies for the trouble. I'm not use to dealing with these kinds of git command issues.

misterjpapa commented:
@jaystarshot  - Thanks for all your efforts here in this PR. Your approval has been over 4 days + 2 days with the weekend ago. Any suggestions on how to get the required code owners to review? Should I make a request in the Dev Slack channel?

steveburnett commented:
> @jaystarshot - Thanks for all your efforts here in this PR. Your approval has been over 4 days + 2 days with the weekend ago. Any suggestions on how to get the required code owners to review? Should I make a request in the Dev Slack channel?

I suggest yes to your question based on the [Pull Request Checklist](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#pull-request-checklist) in CONTRIBUTING.md: 

_If your pull request does not have a reviewer assigned to it after 4 days, [ask for a review in the #dev channel in Slack](https://join.slack.com/t/prestodb/shared_invite/enQtNTQ3NjU2MTYyNDA2LTYyOTg3MzUyMWE1YTI3Njc5YjgxZjNiYTgxODAzYjI5YWMwYWE0MTZjYWFhNGMwNjczYjI3N2JhM2ExMGJlMWM)._

jaystarshot commented:
@tdcmeehan Can you please review?

rschlussel reviewed (approved):
approving based on @jaystarshot's review

jaystarshot commented:
I think we need someone from [team-velox](https://github.com/orgs/prestodb/teams/team-velox) too

tdcmeehan commented on presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:
```diff
@@ -1261,9 +1263,15 @@ protected Session getSession()
     protected Session getMaterializedSession()
     {
         return Session.builder(super.getSession())
+                .setSystemProperty(VERBOSE_OPTIMIZER_INFO_ENABLED, "true")
```
Do we want to include this in the test?  Won't it increase the log size substantially?

jaystarshot replied:
We need this for confirming that the cte was materialized. This doens't increase logging that much since most of the info is added to the session when this property is on

tdcmeehan replied:
Got it, I had forgotten how this property works.
